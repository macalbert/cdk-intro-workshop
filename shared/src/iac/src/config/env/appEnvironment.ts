export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
