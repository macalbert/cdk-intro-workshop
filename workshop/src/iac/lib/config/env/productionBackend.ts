import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
