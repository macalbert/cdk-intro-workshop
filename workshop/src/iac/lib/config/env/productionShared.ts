import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
