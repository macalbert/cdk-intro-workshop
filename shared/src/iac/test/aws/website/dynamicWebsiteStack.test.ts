import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import {
	DynamicWebsiteStack,
	type DynamicWebsiteStackProps,
} from "../../../src/aws/website/dynamicWebsiteStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("DynamicWebsiteStack", () => {
	const env = {
		account: "account",
		region: "us-east-1",
	};

	test("Should_CreateLambdaAndApiGateway_When_StackIsCreated", () => {
		// Arrange
		const stack = new Stack(new App(), "dynamicwebsitestack", { env: env });
		const props = createDynamicWebsiteStackProps(stack);

		// Act
		const actual = new DynamicWebsiteStack(stack, props);

		// Assert
		const template = Template.fromStack(actual).toJSON();

		// Filter out deployment resources which have random IDs
		for (const key in template.Resources) {
			const resource = template.Resources[key];

			if (
				typeof key === "string" &&
				key.startsWith("SsrRestApidynamicwebsitestackDeployment")
			) {
				delete template.Resources[key];
			}
			if (resource.Properties?.DeploymentId?.Ref) {
				resource.Properties.DeploymentId.Ref = undefined;
			}

			// Normalize container image hash in ImageUri to make tests deterministic
			if (resource.Properties?.Code?.ImageUri?.["Fn::Sub"]) {
				// Replace the changing hash with a fixed value for snapshot testing
				resource.Properties.Code.ImageUri["Fn::Sub"] =
					resource.Properties.Code.ImageUri["Fn::Sub"].replace(
						/cdk-hnb659fds-container-assets-account-us-east-1:[a-f0-9]{64}/,
						"cdk-hnb659fds-container-assets-account-us-east-1:HASH-NORMALIZED-FOR-TESTING",
					);
			}
		}

		expect(template).toMatchSnapshot("dynamicwebsitestacktest");
	});

	test("Should_CreateProperLambdaConfiguration_When_StackIsCreated", () => {
		// Arrange
		const stack = new Stack(new App(), "dynamicwebsitestack", { env: env });
		const props = createDynamicWebsiteStackProps(stack);

		// Act
		const actual = new DynamicWebsiteStack(stack, props);

		// Assert
		const template = Template.fromStack(actual);

		// Verify Lambda function configuration
		template.hasResourceProperties("AWS::Lambda::Function", {
			FunctionName: `${props.subdomain}-ssr-${props.envName}`.toLowerCase(),
			MemorySize: props.memorySize,
			Timeout: props.timeoutSeconds,
			Environment: {
				Variables: {
					NODE_ENV:
						props.envName === AppEnvironment.Production
							? "production"
							: "development",
				},
			},
		});

		// Verify API Gateway integration
		template.hasResourceProperties("AWS::ApiGateway::RestApi", {
			Name: `SsrRestApi-${props.stackName}`,
			BinaryMediaTypes: ["*/*"],
		});

		// Verify Route53 record for custom domain
		template.hasResourceProperties("AWS::Route53::RecordSet", {
			Name: `${props.subdomain}.${props.domainName}.`,
			Type: "A",
		});

		// Verify security group created with proper outbound rules
		template.hasResourceProperties("AWS::EC2::SecurityGroup", {
			SecurityGroupEgress: [
				{
					CidrIp: "0.0.0.0/0",
					Description: "Allow HTTPS outbound",
					FromPort: 443,
					IpProtocol: "tcp",
					ToPort: 443,
				},
				{
					CidrIp: "0.0.0.0/0",
					Description: "Allow HTTP outbound",
					FromPort: 80,
					IpProtocol: "tcp",
					ToPort: 80,
				},
			],
		});
	});

	function createDynamicWebsiteStackProps(
		app: Stack,
	): DynamicWebsiteStackProps {
		return {
			env: env,
			subdomain: "website-test",
			name: "DynamicWebsiteStack",
			envName: AppEnvironment.Test,
			githubRepo: "test",
			domainName: "example.com",
			pathDockerFile: "./test/aws/website",
			dockerfile: "Dockerfile",
			certificateArn: `arn:aws:acm:us-east-1:${env.account}:certificate/someid`,
			hostedZoneId: "Z1234567890ABC",
			stackName: "test-dynamic-website",
			vpc: new Vpc(app, "vpc"),
			memorySize: 512,
			timeoutSeconds: 30,
			environmentVariables: {
				TEST_VAR: "test-value",
			},
		};
	}
});
