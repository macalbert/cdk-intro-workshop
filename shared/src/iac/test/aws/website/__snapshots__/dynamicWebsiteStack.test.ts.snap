// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DynamicWebsiteStack Should_CreateLambdaAndApiGateway_When_StackIsCreated: dynamicwebsitestacktest 1`] = `
{
  "Outputs": {
    "SsrRestApitestdynamicwebsiteEndpoint4FFCAB2C": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "SsrRestApitestdynamicwebsiteE322047A",
            },
            ".execute-api.us-east-1.",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "SsrRestApitestdynamicwebsiteDeploymentStageprod088066C6",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ARecordtestdynamicwebsite36E5982E": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "SsrRestApitestdynamicwebsiteCustomDomain6E9CFF83",
              "DistributionDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "SsrRestApitestdynamicwebsiteCustomDomain6E9CFF83",
              "DistributionHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "Z1234567890ABC",
        "Name": "website-test.example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "ApiGatewayAccessLogsFB871B4C": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaRole3A44B857": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaRoleDefaultPolicy75625A82": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            {
              "Action": [
                "ec2:CreateNetworkInterface",
                "ec2:DescribeNetworkInterfaces",
                "ec2:DeleteNetworkInterface",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaRoleDefaultPolicy75625A82",
        "Roles": [
          {
            "Ref": "LambdaRole3A44B857",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SecurityGrouptestdynamicwebsite28AF3389": {
      "Properties": {
        "GroupDescription": "test-dynamic-website-SecurityGroup",
        "GroupName": "test-dynamic-website-SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow HTTPS outbound",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow HTTP outbound",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Fn::ImportValue": "dynamicwebsitestack:ExportsOutputRefvpcA2121C384D1B3CDE",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "SsrLambdaFunction4886C5BE": {
      "DependsOn": [
        "LambdaRoleDefaultPolicy75625A82",
        "LambdaRole3A44B857",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Sub": "account.dkr.ecr.us-east-1.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-account-us-east-1:HASH-NORMALIZED-FOR-TESTING",
          },
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "development",
            "TEST_VAR": "test-value",
          },
        },
        "FunctionName": "website-test-ssr-test",
        "MemorySize": 512,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "LambdaRole3A44B857",
            "Arn",
          ],
        },
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-dynamicwebsitestack-test-stack",
          },
        ],
        "Timeout": 30,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "SecurityGrouptestdynamicwebsite28AF3389",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": [
            {
              "Fn::ImportValue": "dynamicwebsitestack:ExportsOutputRefvpcPrivateSubnet1Subnet934893E8236E2271",
            },
            {
              "Fn::ImportValue": "dynamicwebsitestack:ExportsOutputRefvpcPrivateSubnet2Subnet7031C2BA60DCB1EE",
            },
            {
              "Fn::ImportValue": "dynamicwebsitestack:ExportsOutputRefvpcPrivateSubnet3Subnet985AC459F9514491",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "SsrRestApitestdynamicwebsiteANYApiPermissionTestdynamicwebsitestackm47testdynamicwebsitestackteststackSsrRestApitestdynamicwebsiteEC31DC7FANY61356FCA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SsrLambdaFunction4886C5BE",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:account:",
              {
                "Ref": "SsrRestApitestdynamicwebsiteE322047A",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SsrRestApitestdynamicwebsiteANYApiPermissiondynamicwebsitestackm47testdynamicwebsitestackteststackSsrRestApitestdynamicwebsiteEC31DC7FANYB1D8A301": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SsrLambdaFunction4886C5BE",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:account:",
              {
                "Ref": "SsrRestApitestdynamicwebsiteE322047A",
              },
              "/",
              {
                "Ref": "SsrRestApitestdynamicwebsiteDeploymentStageprod088066C6",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SsrRestApitestdynamicwebsiteANYF08D86F5": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "SsrLambdaFunction4886C5BE",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "SsrRestApitestdynamicwebsiteE322047A",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "SsrRestApitestdynamicwebsiteAccount2643C295": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "SsrRestApitestdynamicwebsiteE322047A",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "SsrRestApitestdynamicwebsiteCloudWatchRole06EDF10F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "SsrRestApitestdynamicwebsiteCloudWatchRole06EDF10F": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-dynamicwebsitestack-test-stack",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "SsrRestApitestdynamicwebsiteCustomDomain6E9CFF83": {
      "Properties": {
        "CertificateArn": "arn:aws:acm:us-east-1:account:certificate/someid",
        "DomainName": "website-test.example.com",
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-dynamicwebsitestack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "SsrRestApitestdynamicwebsiteCustomDomainMapdynamicwebsitestackm47testdynamicwebsitestackteststackSsrRestApitestdynamicwebsiteEC31DC7F9294C551": {
      "Properties": {
        "DomainName": {
          "Ref": "SsrRestApitestdynamicwebsiteCustomDomain6E9CFF83",
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
        "Stage": {
          "Ref": "SsrRestApitestdynamicwebsiteDeploymentStageprod088066C6",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "SsrRestApitestdynamicwebsiteDeployment6BCFA781cadf2ea05a4248c26f5af698d228cf00": {
      "DependsOn": [
        "SsrRestApitestdynamicwebsiteproxyANYD02741BA",
        "SsrRestApitestdynamicwebsiteproxyOPTIONSB1A75B27",
        "SsrRestApitestdynamicwebsiteproxy2CEFAC0B",
        "SsrRestApitestdynamicwebsiteANYF08D86F5",
        "SsrRestApitestdynamicwebsiteOPTIONS93F1B843",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "SsrRestApitestdynamicwebsiteDeploymentStageprod088066C6": {
      "DependsOn": [
        "SsrRestApitestdynamicwebsiteAccount2643C295",
      ],
      "Properties": {
        "AccessLogSetting": {
          "DestinationArn": {
            "Fn::GetAtt": [
              "ApiGatewayAccessLogsFB871B4C",
              "Arn",
            ],
          },
          "Format": "$context.identity.sourceIp $context.identity.caller $context.identity.user [$context.requestTime] "$context.httpMethod $context.resourcePath $context.protocol" $context.status $context.responseLength $context.requestId",
        },
        "DeploymentId": {
          "Ref": undefined,
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
        "StageName": "prod",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-dynamicwebsitestack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "SsrRestApitestdynamicwebsiteE322047A": {
      "Properties": {
        "BinaryMediaTypes": [
          "*/*",
        ],
        "Name": "SsrRestApi-test-dynamic-website",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-dynamicwebsitestack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "SsrRestApitestdynamicwebsiteOPTIONS93F1B843": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Credentials": "'true'",
                "method.response.header.Access-Control-Allow-Headers": "'Cache-Control,Content-Language,Content-Type,Expires,Last-Modified,Pragma,Authorization,X-Amz-Date,X-Api-Key,X-Forwarded-For'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Credentials": true,
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Fn::GetAtt": [
            "SsrRestApitestdynamicwebsiteE322047A",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "SsrRestApitestdynamicwebsiteproxy2CEFAC0B": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "SsrRestApitestdynamicwebsiteE322047A",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "SsrRestApitestdynamicwebsiteproxyANYApiPermissionTestdynamicwebsitestackm47testdynamicwebsitestackteststackSsrRestApitestdynamicwebsiteEC31DC7FANYproxy5BACC110": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SsrLambdaFunction4886C5BE",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:account:",
              {
                "Ref": "SsrRestApitestdynamicwebsiteE322047A",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SsrRestApitestdynamicwebsiteproxyANYApiPermissiondynamicwebsitestackm47testdynamicwebsitestackteststackSsrRestApitestdynamicwebsiteEC31DC7FANYproxy06BA2C0D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "SsrLambdaFunction4886C5BE",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:account:",
              {
                "Ref": "SsrRestApitestdynamicwebsiteE322047A",
              },
              "/",
              {
                "Ref": "SsrRestApitestdynamicwebsiteDeploymentStageprod088066C6",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SsrRestApitestdynamicwebsiteproxyANYD02741BA": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "SsrLambdaFunction4886C5BE",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "SsrRestApitestdynamicwebsiteproxy2CEFAC0B",
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "SsrRestApitestdynamicwebsiteproxyOPTIONSB1A75B27": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Credentials": "'true'",
                "method.response.header.Access-Control-Allow-Headers": "'Cache-Control,Content-Language,Content-Type,Expires,Last-Modified,Pragma,Authorization,X-Amz-Date,X-Api-Key,X-Forwarded-For'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Credentials": true,
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "SsrRestApitestdynamicwebsiteproxy2CEFAC0B",
        },
        "RestApiId": {
          "Ref": "SsrRestApitestdynamicwebsiteE322047A",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
